examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3721 @@
+    //  let state = host.state_mut();
+    ensure!(
+        sender == input.params.owner
+            || host
+                .state()
+                .is_operator(&sender, &input.params.owner, host)
+                .map_or(false, |is_operator| is_operator),
+        ContractError::Unauthorized
+    );
+
+    // Update the state.
+
+    let token_balance = host.state().balance(&TOKEN_ID_WCCD, &input.params.owner, host)?;
+
+    ensure!(token_balance >= input.params.amount, ContractError::InsufficientFunds);
+
+    let set_balance_params = SetBalanceParams {
+        owner:  input.params.owner,
+        amount: input.params.amount,
+        update: Update::Remove,
+    };
+
+    let parameter_bytes = to_bytes(&set_balance_params);
+
+    host.invoke_contract_raw(
+        &state_address,
+        Parameter(&parameter_bytes),
+        EntrypointName::new_unchecked("setBalance"),
+        Amount::zero(),
+    )?;
+
+    let params2: TransferCCDParams = TransferCCDParams {
+        amount: input.params.amount,
+    };
+
+    let parameter_bytes = to_bytes(&params2);
+
+    // Access/Request CCD funds from proxy.
+    host.invoke_contract_raw(
+        &proxy_address,
+        Parameter(&parameter_bytes),
+        EntrypointName::new_unchecked("transferCCD"),
+        Amount::zero(),
+    )?;
+
+    let unwrapped_amount = Amount::from_micro_ccd(input.params.amount.into());
+
+    // Send CCD to receiver.
+    match input.params.receiver {
+        Receiver::Account(address) => host.invoke_transfer(&address, unwrapped_amount)?,
+        Receiver::Contract(address, function) => {
+            host.invoke_contract(
+                &address,
+                &input.params.data,
+                function.as_entrypoint_name(),
+                unwrapped_amount,
+            )?;
+        }
+    }
+
+    // Log the burn event.
+    let parameter_bytes = to_bytes(&BurnEventParams {
+        amount:  input.params.amount,
+        address: input.params.owner,
+    });
+
+    host.invoke_contract_raw(
+        &proxy_address,
+        Parameter(&parameter_bytes),
+        EntrypointName::new_unchecked("logBurnEvent"),
+        Amount::zero(),
+    )?;
+
+    Ok(())
+}
+
+// Contract functions required by CIS2
+
+pub type TransferParameter = TransferParams<ContractTokenId, ContractTokenAmount>;
+
+/// Execute a list of token transfers, in the order of the list.
+///
+/// Logs a `Transfer` event and invokes a receive hook function for every
+/// transfer in the list.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - Any of the transfers fail to be executed, which could be if:
+///     - The `token_id` does not exist.
+///     - The sender is not the owner of the token, or an operator for this
+///       specific `token_id` and `from` address.
+///     - The token is not owned by the `from`.
+/// - Fails to log event.
+/// - Any of the receive hook function calls rejects.
+#[receive(contract = "CIS2-wCCD", name = "transfer", mutable)]
```

Comment:
I don't think this needs to be mutable

Reply:
I get an error removing the `mutable`.

Reply:
ahh right, the invoking here needs the mutability, nevermind then